amr-wind/equation_systems/icns/source_terms/WallMomentumFluxForcing.cpp

```diff
@@ -0,0 +1,214 @@
+    const amrex::Real index = (zref / dx[idir]) - 0.5;
+
+    int kLow = int(std::floor(index));
+    int kHigh = int(std::ceil(index));
+    // if kLow and kHigh point to the same grid cell, separate them by one cell.
+    if (kLow == kHigh) {
+        kHigh += 1;
+    }
+    // if kLow lies below the grid box, bump it up to the first grid level.
+    if (kLow < 0) {
+        kLow = 0;
+        kHigh = kLow + 1;
+    }
+    // if kHigh lies outside the grid box, bump it down to the last grid level.
+    else if (kHigh > bx.bigEnd(idir)) {
+        kHigh = bx.bigEnd(idir);
+        kLow = kHigh - 1;
+    }
+    // kLow = (kLow < 0) ? 0 : kLow;
+    // kHigh = (kLow == kHigh) ? kHigh + 1 : kHigh;
+    // kHigh = (kHigh > bx.bigEnd(idir)) ? bx.bigEnd(idir) : kHigh;
+    // kLow = (kLow == kHigh) ? kLow - 1 : kLow;
+
+    const amrex::Real weightLow = amrex::Real(kHigh) - index;
+    const amrex::Real weightHigh = index - amrex::Real(kLow);
+
+    /*
+        std::cout << "index = " << index << std::endl;
+        std::cout << "kLow = " << kLow << std::endl;
+        std::cout << "kHigh = " << kHigh << std::endl;
+        std::cout << "weightLow = " << weightLow << std::endl;
+        std::cout << "weightHigh = " << weightHigh << std::endl;
+        std::cout << "bx.smallEnd(0) : bx.bigEnd(0) = " << bx.smallEnd(0) << " :
+       " << bx.bigEnd(0) << std::endl; std::cout << "bx.smallEnd(1) :
+       bx.bigEnd(1) = " << bx.smallEnd(1) << " : " << bx.bigEnd(1) << std::endl;
+        std::cout << "bx.smallEnd(2) : bx.bigEnd(2) = " << bx.smallEnd(2) << " :
+       " << bx.bigEnd(2) << std::endl;
+    */
+
+    if (!(bx.smallEnd(idir) == domain.smallEnd(idir))) return;
+    if (idir != 2) return;
+
+    amrex::ParallelFor(
+        amrex::bdryLo(bx, idir),
+        [=] AMREX_GPU_DEVICE(int i, int j, int k) noexcept {
+            // Get the local velocity at the cell center adjacent
+            // to this wall face.
+            const amrex::Real uLow = velocityField(i, j, kLow, 0);
+            const amrex::Real vLow = velocityField(i, j, kLow, 1);
+
+            const amrex::Real uHigh = velocityField(i, j, kHigh, 0);
+            const amrex::Real vHigh = velocityField(i, j, kHigh, 1);
+
+            const amrex::Real u = weightLow * uLow + weightHigh * uHigh;
+            const amrex::Real v = weightLow * vLow + weightHigh * vHigh;
+            const amrex::Real S = std::sqrt((u * u) + (v * v));
+            /*
+                        std::cout << "u = " << uLow << " " << uHigh << " " << u
+               << std::endl; std::cout << "v = " << vLow << " " << vHigh << " "
+               << v << std::endl; std::cout << "S = " << S << std::endl;
+            */
+
+            // Get local tau_wall based on the local conditions and
+            // mean state based on Monin-Obukhov similarity.
+            amrex::Real tau_xz = 0.0;
+            amrex::Real tau_yz = 0.0;
+
+            if (m_wall_shear_stress_type == "constant") {
```

Comment:
These if statements are causing the GPU compilation to fail. Specifically, the m_wall_shear_stress_type is a data member of the class WallMomentumFluxForcing, which would require a capture of 'this' to access. Instead, you should expose these data members before the kernel so that they may be copied onto the GPU; see below example. The second issue will arise in the use of std::string and ==; the compiler will bark about the capture and using host code for the equality. I would advocate for simply using integers to denote different options; sticking with strings would require you to copy to a device vector and then expose the char* and then the equality test becomes nasty.

`int d_wall_shear_stress_type = m_wall_shear_stress_type;
...
amrex::ParallelFor(box,[=] AMREX_GPU_DEVICE(int i, int j, int k) noexcept {
...
if(d_wall_shear_stress_type == 0){
...`